### PR TITLE
Fix `EpochsTFRArray` default drop log initialisation

### LINKED ISSUE #13027: Selecting epochs from `EpochsTFRArray` objects raises obscure error due to incomplete `drop_log` initialisation
### Description of the problem

If you create an `EpochsTFRArray` object with the default `drop_log=None` and then try to select epochs from the object, you get the following error:
image

This is because the `drop_log` is initialised as an empty `tuple`, so there is nothing for `_getitem()` to index:
https://github.com/mne-tools/mne-python/blob/b329515933915fd077495ea41de876119ac04c97/mne/time_frequency/tfr.py#L3108

<br>

This behaviour is in contrast to e.g. `EpochsArray` where `drop_log=None` initialises it as a tuple of empty tuples according to the number of epochs/events (which plays nicely with `_getitem()`):
https://github.com/mne-tools/mne-python/blob/b329515933915fd077495ea41de876119ac04c97/mne/epochs.py#L514-L518

<br>

Working around this is easy (e.g., `drop_log=tuple([() for _ in range(n_epochs)])`), but as a user I would expect the default `drop_log=None` to already work like this.

### Steps to reproduce

```Python
import numpy as np
from mne import create_info
from mne.time_frequency import EpochsTFRArray


n_epochs, n_chans, n_freqs, n_times = (5, 3, 20, 50)
sfreq = 100
data = np.random.rand(n_epochs, n_chans, n_freqs, n_times)
times = np.linspace(0, sfreq/n_times)
freqs = np.arange(1, n_freqs + 1)
info = create_info(n_chans, sfreq, "eeg")
tfr = EpochsTFRArray(info=info, data=data, times=times, freqs=freqs, drop_log=None)
tfr[0]
```

### Link to data

_No response_

### Expected results

`tfr[0]` should give the TFR data for the first epoch.

### Actual results

image

### Additional information

Using dev version.

### PR DESCRIPTION
**Pull request by tsbinns**
#### Reference issue (if any)

Fixes #13027

#### What does this implement/fix?

Expands the logic for `EpochsTFRArray` default `drop_log` initialisation so `__getitem__()` works with default params.

### CODE REVIEW

File: mne/time_frequency/tests/test_tfr.py [hunk 1]
```diff
@@ -1218,25 +1218,29 @@ def test_averaging_freqsandtimes_epochsTFR():
         avgpower = power.average(method=lambda x: np.mean(x, axis=2), **kwargs)
 
 
-@pytest.mark.parametrize("n_drop", (0, 2))
-def test_epochstfr_getitem(epochs_full, n_drop):
+@pytest.mark.parametrize("n_drop, as_array", ((0, False), (0, True), (2, False)))
```
Comment (drammock) [inline mne/time_frequency/tests/test_tfr.py:1221]: can we change this to `as_tfr_array` or even `as_tfrarray`?  `as_array` has strong associations with https://numpy.org/doc/stable/reference/generated/numpy.asarray.html so I went down the wrong mental path reading this diff.

File: mne/time_frequency/tfr.py [hunk 2]
```diff
@@ -3105,8 +3105,14 @@ def __setstate__(self, state):
         ).squeeze(axis=0)
         self.events = state.get("events", _ensure_events(fake_events))
         self.event_id = state.get("event_id", _check_event_id(None, self.events))
-        self.drop_log = state.get("drop_log", tuple())
         self.selection = state.get("selection", np.arange(n_epochs))
+        self.drop_log = state.get(
+            "drop_log",
+            tuple(
+                () if k in self.selection else ("IGNORED",)
+                for k in range(max(len(self.events), max(self.selection) + 1))
```
Comment (drammock) [inline mne/time_frequency/tfr.py:3113]: scratching my head about this line. When would we *not* want just `range(len(self.events))`?

Comment (larsoner) [reply mne/time_frequency/tfr.py:3113]: We always have `len(self.events)` == `len(epochs)`. But let's say you originally had 100 events and just selected the last 10 on way or another, that resulting `Epochs` object shouldhave for example
```
len(epochs) == 10
len(epochs.selection) == 10
len(epochs.events) == 10
list(epochs.selection) == list(range(90, 100))
```
and in general for anything not in `epochs.selection` you should be able to query that index in `epochs.drop_log` to see why it's not part of `epochs`. So `epochs.drop_log[30]` for example should exist and be non-empty meaning it was dropped for a reason, whereas `epochs.drop_log[90:]` should exist and all be empty (i.e., those are the epochs that have been kept).

Comment (drammock) [reply mne/time_frequency/tfr.py:3113]: right, I forgot about events getting truncated/sliced/subselected.

Am I right then that we don't need to care about the case of having 30 events, and selecting only the *middle* 10?  IIUC, if that happened via normal means, the `drop_log` would be present and the code in this diff won't be reached. It's only for EpochsTFRArray (where we don't know the dropping history) where we have to spawn a "fake" drop log, and in that case we can't know if (say) the last 10 epochs were dropped, but we also don't *need* to know that for things to work correctly.

Comment (larsoner) [reply mne/time_frequency/tfr.py:3113]: Yeah I think we just need some drop log that satisfies the necessary expectations about length of the drop log, selection values, and number of events.

Comment (tsbinns) [reply mne/time_frequency/tfr.py:3113]: So something like
```
tuple(() if k in self.selection else ("IGNORED",) for k in len(self.events))
```
to account for the fact that a non-default `selection` param could be passed?

Comment (larsoner) [reply mne/time_frequency/tfr.py:3113]: No the `len(self.events)` won't be enough, I think what you have here with the `max(...)` is more correct

Comment (larsoner) [reply mne/time_frequency/tfr.py:3113]: ... or you have to make sure `self.selection = np.arange(len(self.events))`

Comment (tsbinns) [reply mne/time_frequency/tfr.py:3113]: Ah sorry. Yeah at the moment whatever selection can be passed so `self.selection = np.arange(len(self.events))` can't be assumed.

File: mne/time_frequency/tests/test_tfr.py (hunk 1 above)
Comment (tsbinns) [reply mne/time_frequency/tests/test_tfr.py]: Have updated this now.

**Comment by tsbinns** [comment]
Have now switched from `as_array` to `as_tfr_array` in the test, and have included a changelog entry.

**Comment by larsoner** [comment]
Thanks @tsbinns !

**Pull request merged**